Replicators (#18012)
canonical/lxd: doc/explanation/replicators.md

---
myst:
  html_meta:
    description: An explanation of LXD replicators and how they enable active-passive disaster recovery across cluster links.
---

(exp-replicators)=
# Replicators

Replicators are LXD entities that periodically copy instances from one cluster to another across a {ref}`cluster link <exp-cluster-links>`. They are designed for active-passive disaster recovery, where a leader cluster runs all workloads and a standby cluster stays ready to take over if the leader fails.

(exp-replicators-concepts)=
## Leader and standby projects

> Replication is configured at the project level. Each participating cluster has a project with the same name, and each project is assigned a `replica.mode`:

Reviewer: Is it "each cluster" or "both clusters"? "Each" makes it sound like there can be multiple (more than two) clusters that participate. Can there be, for example, a leader cluster + more than one standby cluster both linked to the leader?

Author: I've switched the wording to "both clusters" - replication is a two-cluster relationship.

> Replication is configured at the project level. Each participating cluster has a project with the same name, and each project is assigned a `replica.mode`:

Reviewer: ```suggestion
Replication is configured at the project level. Each participating cluster has a project with the same name, and each project is assigned a {config:option}`project-replica:replica.mode`:
```

- `leader`: The project is writable. Instances in this project are the source of replication. The replicator runs from this cluster.

> - `standby`: The project is read-only. Instances in this project are replicas, kept in sync by the replicator. Instances cannot be started or modified directly until the project is promoted to `leader` during a failover.

Reviewer: Does this mean that the standby project is read-only to users, but can be written to by the leader project? Since the leader project is able to copy instances to the standby project.

Author: Correct.

Author: I've clarified this part - we only disallow instance creation by users.

The leader project pushes its instances to the standby project over the cluster link. The standby project mirrors the leader at the time of the last replicator run.

(exp-replicators-how)=
## How replication works

When a replicator runs, LXD performs an incremental refresh of every instance in the leader project to the standby project. Instances that do not yet exist on the standby are created; existing instances are updated to match the leader's current state.

> If `snapshot=true` is configured on the replicator, LXD creates a point-in-time snapshot of each instance on the leader before the refresh. This provides a consistent rollback point on the source cluster in case anything goes wrong during replication.

Reviewer: ```suggestion
If {config:option}`replicator-conf:snapshot` is set to `true` on the replicator, LXD creates a point-in-time snapshot of each instance on the leader before the refresh. This provides a consistent rollback point on the source cluster in case anything goes wrong during replication.
```

> Replication can be triggered manually with `lxc replicator run`, or scheduled automatically using a cron expression in the `schedule` configuration key.

Reviewer: ```suggestion
Replication can be triggered manually with `lxc replicator run`, or scheduled automatically using a cron expression in the {config:option}`replicator-conf:schedule` configuration key.
```

(exp-replicators-failover)=
## Failover and recovery

> If the leader cluster fails, the standby project can be promoted by setting `replica.mode=leader` on the standby cluster. This makes the project writable and allows instances to be started.

Reviewer: Is this always a manual operation?

Author: Yes, at this time, failover is a manual process.

When the original leader comes back online, it can be re-synced from the new leader by running the replicator in restore mode (`lxc replicator run --restore`), then returning both projects to their original roles.

See {ref}`howto-replicators-dr` for step-by-step instructions.

(exp-replicators-vs-storage-replication)=
## Replicators vs. storage replication

LXD supports two distinct approaches to cross-site disaster recovery:

| | Replicators | Storage replication |
|---|---|---|
| **Level** | LXD instance layer | Storage array layer |
| **Mechanism** | Incremental instance refresh over cluster links | Vendor storage replication (Ceph RBD mirroring, PowerFlex RCG, etc.) |

> | **Scheduling** | Controlled by LXD (`schedule` config key) | Controlled by the storage vendor |

Reviewer: ```suggestion
| **Scheduling** | Controlled by LXD ({config:option}`replicator-conf:schedule` config key) | Controlled by the storage vendor |
```

| **Requires cluster link** | Yes | No |
| **Recovery method** | Promote standby project with `replica.mode=leader` | Promote storage array, then run `lxd recover` |
| **Snapshot support** | Optional pre-replication snapshots | Depends on storage vendor |

> Use replicators when you want LXD to manage replication end-to-end across two clusters without dependency on a specific storage backend. Use storage replication when you need replication at the storage array level, or when you are not using cluster links.

Reviewer: ```suggestion
Use replicators when you want LXD to manage replication end-to-end across two clusters without dependency on a specific storage backend. Use {ref}`storage replication <disaster-recovery-replication>` when you need replication at the storage array level, or when you are not using cluster links.
```
